tests/PHPStan/Analyser/data/bug-6329.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+namespace Bug6329;
+
+use function PHPStan\Testing\assertType;
+
+/**
+ * @param mixed $a
+ */
+function nonEmptyString1($a): void
+{
+	if (is_string($a) && '' !== $a || null === $a) {
+		assertType('non-empty-string|null', $a);
+	}
+
+	if ('' !== $a && is_string($a) || null === $a) {
+		assertType('non-empty-string|null', $a);
+	}
+
+	if (null === $a || is_string($a) && '' !== $a) {
+		assertType('non-empty-string|null', $a);
+	}
+
+	if (null === $a || '' !== $a && is_string($a)) {
+		assertType('non-empty-string|null', $a);
+	}
+}
+
+/**
+ * @param mixed $a
+ */
+function nonEmptyString2($a): void
+{
+	if (is_string($a) && strlen($a) > 0 || null === $a) {
+		assertType('non-empty-string|null', $a);
+	}
+
+	if (null === $a || is_string($a) && strlen($a) > 0) {
+		assertType('non-empty-string|null', $a);
+	}
+}
+
+
+/**
+ * @param mixed $a
+ */
+function int1($a): void
+{
+	if (is_int($a) && 0 !== $a || null === $a) {
+		assertType('int<min, -1>|int<1, max>|null', $a);
+	}
+
+	if (0 !== $a && is_int($a) || null === $a) {
+		assertType('int<min, -1>|int<1, max>|null', $a);
+	}
+
+	if (null === $a || is_int($a) && 0 !== $a) {
+		assertType('int<min, -1>|int<1, max>|null', $a);
+	}
+
+	if (null === $a || 0 !== $a && is_int($a)) {
+		assertType('int<min, -1>|int<1, max>|null', $a);
+	}
+}
+
+/**
+ * @param mixed $a
+ */
+function int2($a): void
+{
+	if (is_int($a) && $a > 0 || null === $a) {
+		assertType('int<1, max>|null', $a);
+	}
+
+	if (null === $a || is_int($a) && $a > 0) {
+		assertType('int<1, max>|null', $a);
+	}
+}
+
+
+/**
+ * @param mixed $a
+ */
+function true($a): void
+{
+	if (is_bool($a) && false !== $a || null === $a) {
+		assertType('true|null', $a);
+	}
+
+	if (false !== $a && is_bool($a) || null === $a) {
+		assertType('true|null', $a);
+	}
+
+	if (null === $a || is_bool($a) && false !== $a) {
+		assertType('true|null', $a);
+	}
+
+	if (null === $a || false !== $a && is_bool($a)) {
+		assertType('true|null', $a);
+	}
+}
+
+/**
+ * @param mixed $a
+ */
+function nonEmptyArray1($a): void
+{
+	if (is_array($a) && [] !== $a || null === $a) {
```

Comment:
This makes me think whether this also works for e.g.
`is_array($a) && [] !== $a || true === $a`
`is_array($a) && [] !== $a || false === $a`
`is_array($a) && [] !== $a || “abc“ === $a`
`is_array($a) && [] !== $a || 123 === $a`

But lets wait for ondrejs feedback first. The PR as is seems to add already a lot of great type improvements

Reply:
I did a quick test, all of those cases are working fine. Which is not surprising I guess :)
Results are
```
non-empty-array|true
non-empty-array|false
'abc'|non-empty-array
123|non-empty-array
```
I did not check though if those were already working before or not :)